### PR TITLE
feat(em): add overvote to hmpb deck

### LINKED ISSUE #1711: Augment VxAdmin test decks with pre-marked overvoted HMBP
Augment VxAdmin test decks with pre-marked overvoted HMBP

### PR DESCRIPTION
**Pull request by adghayes**
## Overview
Closes #1711. Appends an overvote ballot to the HMPB test deck printer from VxAdmin for L&A. 

## Demo Video or Screenshot
Screen Shot 2022-05-12 at 3 16 11 PM

## Testing Plan 

Updated snapshot. @arsalansufi it's worth noting that the snapshots used for HMPB ballots are mocks that include the precinct, election, and ballot style, but don't actually include the votes. Same as in #1808. So we are testing that an additional ballot of the correct ballot style is being added, but not that it is properly marked. 

## Checklist
- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.
- [ ] I have added JSDoc comments to any newly introduced exports

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: frontends/election-manager/src/utils/election.ts [hunk 1]
```diff
@@ -264,3 +264,45 @@ export function generateBlankBallots({
...
+  const votes: VotesDict = {};
+
+  for (const ballotStyle of precinctBallotStyles) {
+    const contests = election.contests.filter(
+      (c) =>
+        ballotStyle.districts.includes(c.districtId) &&
+        ballotStyle.partyId === c.partyId
+    );
+
+    for (const contest of contests) {
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.ts:289]: While it's reasonably clear reading the code that we're generating a minimally overvoted ballot, worth adding a comment noting that! Maybe with a (super brief) explanation as to why we made that call

File: frontends/election-manager/src/utils/election.ts [hunk 2]
```diff
@@ -264,3 +264,45 @@ export function generateBlankBallots({
...
+        votes[contest.id] = contest.candidates.slice(0, contest.seats + 1);
+      } else {
+        continue;
+      }
+
+      ballot.ballotStyleId = ballotStyle.id;
+      ballot.votes = votes;
+      return ballot;
+    }
+  }
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.ts:307]: > Updated snapshot. @arsalansufi it's worth noting that the snapshots used for HMPB ballots are mocks that include the precinct, election, and ballot style, but don't actually include the votes. Same as in https://github.com/votingworks/vxsuite/pull/1808. So we are testing that an additional ballot of the correct ballot style is being added, but not that it is properly marked.

Ooh good to know! Let's test this function through additional means then. A set of unit tests targeted at this helper, covering the various contest cases as well as edge cases (no contest to validly overvote, etc.), could work well!

In some of our other frontend packages (frontends/bmd), we have automated code coverage checks that verify that every branch of code is covered by a test. While we don't have that set up in frontends/election-manager, I still think the "cover every branch" mindset is helpful, in this case writing tests that cover all the branches within this helper

File: frontends/election-manager/src/utils/election.ts [hunk 3]
```diff
@@ -264,3 +264,45 @@ export function generateBlankBallots({
...
+        contest.candidates.length > contest.seats
+      ) {
+        votes[contest.id] = contest.candidates.slice(0, contest.seats + 1);
+      } else {
+        continue;
+      }
+
+      ballot.ballotStyleId = ballotStyle.id;
+      ballot.votes = votes;
+      return ballot;
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.ts:305]: Echoing back my understanding of the code: If there's no contest to validly overvote, I _think_ we'll end up returning the last ballot style with no votes - is that correct / what we want? Based on discussion in https://github.com/votingworks/vxsuite/issues/1711, I believe we want to return nothing at all in that case

**Review by arsalansufi — COMMENTED** [review]
Thanks again for driving the discussion in https://github.com/votingworks/vxsuite/issues/1711 - it's given me lots of clarity too! :smile:

File: frontends/election-manager/src/utils/election.ts [hunk 4]
```diff
@@ -264,3 +264,45 @@ export function generateBlankBallots({
...
   return ballots;
 }
+
+export function generateOvervoteBallot({
+  election,
+  precinctId,
+}: {
+  election: Election;
+  precinctId: PrecinctId;
+}): Dictionary<string | VotesDict> | undefined {
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.ts:274]: Worth adding a comment above the function noting when we expect to return `undefined` (I believe when there's no contest to validly overvote)

File: frontends/election-manager/src/utils/election.test.ts [hunk 5]
```diff
@@ -16,3 +18,77 @@ test('getBallotPath allows digits in file names', () => {
     'election-d34db33f-precinct-north-springfield-id-21-style-77-English-live-absentee.pdf'
   );
 });
+
+describe('generateOvervoteBallot adds an overvote conflict when possible', () => {
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.test.ts:22]: Very minor: People often like to have the combination of their `describe` and `test` / `it` commands read with continuity. Shortening this `describe` to `describe('generateOvervoteBallot', () => {` would accomplish that:

```
generateOvervoteBallot
  minimally overvotes an initial candidate contest
  etc.
```

I think the `adds an overvote conflict when possible` is actually already implied through your descriptive `test` strings!

File: frontends/election-manager/src/utils/election.test.ts [hunk 6]
```diff
@@ -16,3 +18,77 @@ test('getBallotPath allows digits in file names', () => {
...
+
+    expect(Object.keys(votes)).toEqual([overvoteContestId]);
+    expect(votes[overvoteContestId]?.length).toEqual(2);
+  });
+
+  test('overvotes an initial yes-no contest', () => {
+    const election = _.cloneDeep(electionSample);
+
+    // remove all but the yes-no contests
+    _.remove(election.contests, (c) => c.type !== 'yesno');
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.test.ts:43]: While I know we reached the consensus that lodash is fine for us to use in these tests, especially for `cloneDeep`, it's still nice to avoid using it when easily possible. This `remove` can be easily replaced with a `filter`:

```ts
election.contests = election.contests.filter(c => c.type === 'yesno')
```

Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.test.ts:43]: While I know we reached the consensus that lodash is fine for us to use in these tests, especially for `cloneDeep`, it's still nice to avoid using it when easily possible. This `remove` can be easily replaced with a `filter`:

```ts
election.contests = election.contests.filter(c => c.type === 'yesno')
```

File: frontends/election-manager/src/utils/election.test.ts [hunk 7]
```diff
@@ -16,3 +18,77 @@ test('getBallotPath allows digits in file names', () => {
...
+      election: electionSample,
+      precinctId,
+    });
+    expect(overvoteBallot).toBeDefined();
+
+    const votes = overvoteBallot!.votes as VotesDict;
+    const overvoteContestId = electionSample.contests[0].id;
+
+    expect(Object.keys(votes)).toEqual([overvoteContestId]);
+    expect(votes[overvoteContestId]?.length).toEqual(2);
```
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.test.ts:36]: Did you consider boiling this down to a single assertion of what's in votes?

```test
expect(votes).toEqual(/* Exactly what's in votes */)
```

I can see arguments in favor of both approaches! The current approach makes the qualities of the state you want clear (e.g. 2 votes). The only reason I lean toward the single exact assertion is that the current code could pass if we overvoted a yes-no contest instead of a candidate contest since `['yes', 'no']` is also of length `2`

Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.test.ts:36]: Did you consider boiling this down to a single assertion of what's in votes?

```test
expect(votes).toEqual(/* Exactly what's in votes */)
```

I can see arguments in favor of both approaches! The current approach makes the qualities of the state you want clear (e.g. 2 votes). The only reason I lean toward the single exact assertion is that the current code could pass if we overvoted a yes-no contest instead of a candidate contest since `['yes', 'no']` is also of length `2`

**Review by arsalansufi — APPROVED** [review]
This looks great! Thanks for iterating on the tests!

File: frontends/election-manager/src/utils/election.test.ts (hunk 5 above)
Comment (arsalansufi) [inline frontends/election-manager/src/utils/election.test.ts:22]: Very minor: People often like to have the combination of their `describe` and `test` / `it` commands read with continuity. Shortening this `describe` to `describe('generateOvervoteBallot', () => {` would accomplish that:

```
generateOvervoteBallot
  minimally overvotes an initial candidate contest
  etc.
```

I think the `adds an overvote conflict when possible` is actually already implied through your descriptive `test` strings!

File: frontends/election-manager/src/utils/election.test.ts (hunk 6 above)
Comment (adghayes) [reply frontends/election-manager/src/utils/election.test.ts:43]: The problem I encountered is that the properties on the `Election` type are all `readonly`, so I can't change the array reference. I had to filter the arrays in place, which brought me back to lodash. If there's another way around the `readonly` flag, I would do that.

File: frontends/election-manager/src/utils/election.test.ts (hunk 7 above)
Comment (adghayes) [reply frontends/election-manager/src/utils/election.test.ts:36]: Good catch! While the current approach might be a little easier to read without knowing the fixture data, I think I should simplify as you suggest to avoid passing a yes/no situation.

File: frontends/election-manager/src/utils/election.test.ts (hunk 5 above)
Comment (adghayes) [reply frontends/election-manager/src/utils/election.test.ts:22]: Cool, I'll edit the descriptions in the next commit.

File: frontends/election-manager/src/utils/election.test.ts (hunk 6 above)
Comment (arsalansufi) [reply frontends/election-manager/src/utils/election.test.ts:43]: Ah got it! In that case, this seems totally reasonable

**Review by arsalansufi — APPROVED** [review]
Updates LGTM as well!

File: frontends/election-manager/src/utils/election.test.ts (hunk 6 above)
Comment (arsalansufi) [reply frontends/election-manager/src/utils/election.test.ts:43]: Retroactive thought: We actually can avoid lodash `cloneDeep` by using object and array spreads! https://github.com/votingworks/vxsuite/pull/1890/commits/fe843ac4dd7bb860dfdbe410baad8671d20992a8
